### PR TITLE
Duplicate task runner instances before usage in engine

### PR DESCRIPTION
**Pull request by madkinsz**
Allows concurrent execution of the same flow, avoiding the "task runner is already started" issue by creating a new instance for each run. Avoids introducing new types or patterns — task runner implementations will need to implement the `duplicate` method if they take options. We use this instead of `copy` to avoid including any runtime state.

Closes https://github.com/PrefectHQ/prefect/issues/7319
Closes https://github.com/PrefectHQ/prefect/issues/8307

On release, Dask/Ray will ignore options passed to the task runner until their implementations are released:

Dask implementation https://github.com/PrefectHQ/prefect-dask/pull/93
Ray implementation https://github.com/PrefectHQ/prefect-ray/pull/83

We could make this an abstract method if we'd rather they just crashed and burned.

## Example

Concurrent instances of the same subflow

```python
from prefect import flow, task
import asyncio


@task
async def test_task():
    return 1


@flow(log_prints=True)
async def my_child(i):
    assert await test_task() == 1
    print(i)


@flow
async def my_parent():
    coros = [my_child(i) for i in range(10)]
    await asyncio.gather(*coros)


asyncio.run(my_parent())
```

Subflows sharing a dynamically created Dask cluster

```python
from prefect import flow, task
import asyncio
from prefect_dask import DaskTaskRunner
from prefect.context import get_run_context


@task
async def test_task():
    return 1


@flow(log_prints=True)
async def my_child(i):
    assert await test_task() == 1
    print(i)


@flow(task_runner=DaskTaskRunner())
async def my_parent():
    child_task_runner = DaskTaskRunner(
        address=get_run_context().task_runner._client.scheduler_info()["address"]
    )
    coros = [my_child.with_options(task_runner=child_task_runner)(i) for i in range(5)]
    await asyncio.gather(*coros)


if __name__ == "__main__":
    asyncio.run(my_parent())
```

### CODE REVIEW

File: src/prefect/task_runners.py [hunk 1]
```diff
@@ -109,6 +109,21 @@ def concurrency_type(self) -> TaskConcurrencyType:
...
+        return type(self)()
+
+    def __eq__(self, __value: object) -> bool:
+        """
+        Returns true if the task runners use the same options.
+        """
+        if type(__value) == type(self):
+            return True
+        else:
+            return NotImplemented
```
Comment (rsampaths16) [inline src/prefect/task_runners.py:125]: question: is the intention here to let the other type determine equality?

**Review by rsampaths16 — COMMENTED** [review]
Hi @madkinsz the approach seems good - have added my thoughts on it

File: src/prefect/task_runners.py [hunk 2]
```diff
@@ -109,6 +109,21 @@ def concurrency_type(self) -> TaskConcurrencyType:
     def name(self):
         return type(self).__name__.lower().replace("taskrunner", "")
 
+    def duplicate(self):
+        """
+        Return a new task runner instance with the same options.
+        """
+        return type(self)()
+
+    def __eq__(self, __value: object) -> bool:
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/task_runners.py:118]: Fair! This was the auto-completed signature from VSCode and I just went with it :) I like `other` more.

File: tests/test_flows.py [hunk 3]
```diff
@@ -752,6 +753,23 @@ def parent(x, y=2, z=3):
 
         assert parent(1, 2) == 6
 
+    async def test_concurrent_async_subflow(self):
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply tests/test_flows.py:756]: 👍 good idea

File: tests/test_flows.py [hunk 4]
```diff
@@ -752,6 +753,23 @@ def parent(x, y=2, z=3):
...
+        @task
+        async def test_task():
+            return 1
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply tests/test_flows.py:759]: There are lots of async operations in orchestration of the flows that yields. We could add an `asyncio.sleep(0)` here if we want to yield, but I don't want to slow down the test

**Review by desertaxle — APPROVED** [review]
LGTM!

I lean slightly towards `duplicate` being an abstract method to make failures related to `duplicate` in `prefect-dask` and `prefect-ray` easier to debug, but we can cut new releases for those very quickly after 2.10.6 is released, so I think the difference is negligible.

File: src/prefect/task_runners.py (hunk 1 above)
Comment (madkinsz) [reply src/prefect/task_runners.py:125]: Yep exactly this signals to Python that it should call __eq__ on the other type.

**Comment by madkinsz** [comment]
@desertaxle another option is to have the base class `__init__` capture `**kwargs` then have `duplicate` use those kwargs. We'd then need to update the implementations to pass all of their input arguments up to the `super().__init__()` call though so it's kind of the same problem.

I'm not sure there's a great way to detect a lack of implementation here without the abstract method though.

**Comment by madkinsz** [comment]
I've added https://github.com/PrefectHQ/prefect/pull/9342/commits/d5443bb4cdb3529c7ccd7ee50b7060811670e1f6 which will at least make a task runner fail the standard test suite without implementing this

**Comment by madkinsz** [comment]
@desertaxle I got uncomfortable with the breaking change and added https://github.com/PrefectHQ/prefect/pull/9342/commits/4c045b2e493f0dd15e26c8c5bdd5f18fb5157e64

**Comment by desertaxle** [comment]
> @desertaxle I got uncomfortable with the breaking change and added [4c045b2](https://github.com/PrefectHQ/prefect/commit/4c045b2e493f0dd15e26c8c5bdd5f18fb5157e64)

Nice, that looks solid!

File: tests/test_flows.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rsampaths16) [reply tests/test_flows.py:759]: Yeah `asyncio.sleep(0)` should work. Verified from the documentation: https://docs.python.org/3.11/library/asyncio-task.html#sleeping

**Pull request merged**